### ISSUE TITLE
0.5.0: poll EMR logs for counters

### ISSUE THREAD

**Issue by andrey-cliqz** [Yelp/mrjob#1292]
Currently if mrjob can't fetch logs via SSH, it goes to S3 for them. When it does this fallback, it waits for 10 minutes for the logs to be fully synced (this delay is hard-coded in [_S3_LOG_WAIT_MINUTES](https://github.com/Yelp/mrjob/blob/master/mrjob/emr.py#L163). The problem is, mrjob makes this delay for every step of the finished job, e.g. if we have a job with 2 steps we would wait 20 minutes and for 6 step job we would wait 1 hour. This long delay kills the idea of pooling clusters because while we wait for multi-step job counters, the cluster usually gets killed.

I made a bit hacky workaround for this behaviour (removed some code for brevity):
```
import time
from mrjob import emr

DELAY_BETWEEN_ATTEMPTS = 30  # seconds
emr._S3_LOG_WAIT_MINUTES = 0  # mrjob 0.5.0 has this value hard-coded to 10, which is not what we want

def fetch_emr_counters(runner, attempts=10):
    try:
        counters = runner.counters()
        if all(counters):
            return counters

        if attempts and not all(counters):
            time.sleep(DELAY_BETWEEN_ATTEMPTS)
            runner._wait_for_steps_to_complete()

            return fetch_emr_counters(runner, attempts-1)
        else:
            return counters
    except Exception:
        return []
```

I was just thinking if it is worth having similar behaviour by default (or maybe provide a separate public method to do the log polling)?

**Comment by davidmarin** [comment]
Yeah, it really should be waiting until 10 minutes after the step finished (which may be less than 10 minutes, or no time at all, for steps that have already completed). I'll try to address this in 0.5.1.

I'm unclear how the 10 minute delay affects pooling; mrjob will submit all your steps to EMR at the same time, and once the (pooled) cluster runs them, other `EMRJobRunner`s will happily submit steps to it; they don't know or care that the runner for your particular job is twiddling its thumbs waiting for logs to upload. Or am I missing something?

**Comment by andrey-cliqz** [comment]
You are right, when the step is finished and the current runner is waiting its 10 minute delay, another runner can submit a new job. It works fine until you try to run jobs one after another, e.g. like this:
```
for job_class in [Job1, Job2, ...]:
    args = get_job_args(job_class)
    with job_class(args).make_runner() as runner:
        runner.run()
        if not counters_are_ok(runner.counters()):
            break
```
The code above is a simplified version of the approach we use (we use MRJob with [Luigi](http://luigi.readthedocs.org/en/stable/)). We check that counters from a task are good to decide whether we can run the next one. I agree that our case may be special, but I think others could also benefit from the log polling approach.

**Comment by davidmarin** [comment]
Ah, I see, you want to get the counters just as soon as they're available.

The problem is that I don't think there's a way for mrjob to distinguish between partial and complete logs. I don't want to create edge cases where mrjob parses some of the counters and silently ignores others.

Is there something that's keeping you from setting up SSH tunneling? That gets the counters fast.

**Comment by andrey-cliqz** [comment]
Thanks for the suggestion, I'll try ssh tunnelling, but this polling approach has worked quite well for me in the last few months and I never had partial counters (I had a similar workaround for mrjob 0.4.6 too). 

I have also one other suggestion related to counters: it would be convenient if `runner.counters()` returned counter dictionaries in the same order the steps are organised on the job. Currently I see that they are returned in the reversed order so I had to patch [step_ids_for_job()](https://github.com/Yelp/mrjob/blob/master/mrjob/emr.py#L224) like this:
```
def step_ids_for_job(steps, job_key):
    step_ids = {}
    for step in steps:
        if step.name.startswith(job_key):
            job_step_num = STEP_NAME_RE.match(step.name).group(1)
            step_ids[int(job_step_num)] = step.id
    return [step_ids[job_step_num] for job_step_num in sorted(step_ids)]
```
Probably EMR API behaves a bit differently for different accounts, or maybe there is some other reason not to change this, but getting the counters in the reversed order doesn't seem logical.

**Comment by davidmarin** [comment]
Oh, I see what you mean, [ListSteps](http://docs.aws.amazon.com/ElasticMapReduce/latest/API/API_ListSteps.html) returns the most recently queued step first (presumably because if you had to paginate, the most recent steps are the most relevant).

Sorry, `counters()` has always been in the order steps are executed. Works that way for the `hadoop`, `local`, and `inline` runners too.

**Comment by davidmarin** [comment]
Can you tell me how mrjob 0.5.2 works for you? I'm guessing the fix for #1316 mitigates this somewhat.
